fea-rs/src/compile/glyphsapp_syntax_ext.rs

```diff
@@ -0,0 +1,222 @@
+            typed::GlyphsAppOperator::Div(_) => Operator::Div,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::TokenSet;
+
+    use self::typed::AstNode;
+
+    use super::*;
+
+    fn parse_expr(text: &str) -> typed::GlyphsAppNumberExpr {
+        let node = crate::parse::parse_node(text, |p| {
+            crate::parse::grammar::expect_glyphs_number_value(p, TokenSet::EMPTY);
+        });
+        let node = typed::GlyphsAppNumber::cast(&node.into()).unwrap();
+        node.iter()
+            .find_map(typed::GlyphsAppNumberExpr::cast)
+            .unwrap()
+    }
+
+    fn simple_number_value() -> HashMap<NormalizedLocation, f64> {
+        let loc1 = NormalizedLocation::for_pos(&[("wght", 0.1)]);
+        let loc2 = NormalizedLocation::for_pos(&[("wght", 0.9)]);
+        HashMap::from([(loc1, 10.), (loc2, 23.)])
+    }
+
+    // a little helper for more readable assertions
+    fn ordered_eq<const N: usize>(lhs: &ResolvedValue, rhs: [i16; N]) -> bool {
+        let ResolvedValue::Variable(lhs) = lhs else {
+            return false;
+        };
+        let mut ordered: Vec<_> = lhs.iter().collect();
+        ordered.sort_by_key(|(k, _)| *k);
+        let ordered = ordered.iter().map(|(_, v)| **v).collect::<Vec<_>>();
+        ordered == rhs
+    }
+
+    #[test]
+    fn scalar_arithmetic() {
+        let text = "${2 + 4 * 5 / 7 - 3}";
+
+        let expr = parse_expr(text);
+        let val = resolve_glyphs_app_expr(&expr, |_| Default::default());
+        let expected = (2f64 + 4. * 5. / 7. - 3.).round() as i16;
+        assert_eq!(val, ResolvedValue::Scalar(expected));
+    }
+
+    #[test]
+    fn mixed_arithmetic() {
+        let text = "${padding * 2}";
+        let expr = parse_expr(text);
+        let val: ResolvedValue = resolve_glyphs_app_expr(&expr, |_| simple_number_value());
+        assert!(ordered_eq(&val, [20i16, 46]));
+    }
+
+    #[test]
+    fn mixed_arithmetic_other_order() {
+        let text = "${2 * padding}";
+        let expr = parse_expr(text);
+        let val: ResolvedValue = resolve_glyphs_app_expr(&expr, |_| simple_number_value());
+        assert!(ordered_eq(&val, [20i16, 46]));
+    }
+
+    #[test]
+    fn negative_numbers() {
+        let text = "${2 - -3}";
+        let expr = parse_expr(text);
+        let val = resolve_glyphs_app_expr(&expr, |_| Default::default());
+        assert_eq!(val, ResolvedValue::Scalar(2 + 3));
+    }
+
+    #[test]
+    fn negative_parse_ambiguity() {
+        let text = "${2-3}";
+        let expr = parse_expr(text);
+        let val = resolve_glyphs_app_expr(&expr, |_| Default::default());
+        assert_eq!(val, ResolvedValue::Scalar(2 - 3));
+    }
+
+    #[test]
+    fn float_math() {
+        let text = "${padding/2.4}";
+        let expr = parse_expr(text);
+        let val: ResolvedValue = resolve_glyphs_app_expr(&expr, |_| simple_number_value());
+        assert!(ordered_eq(&val, [4i16, 10]));
+    }
+}
```

Comment:
I don't seem to see any tests for invalid states, malformed input, etc producing errors?

Reply:
those things are part of the parse tests; this is just testing that we are actually evaluating valid expressions as expected.